### PR TITLE
Add deployment pattern recommendations

### PR DESCRIPTION
**Pull request by aefhm**
## Description

Add technical notes on proxies.

Relates to https://github.com/oasisprotocol/docs/pull/818.

### CODE REVIEW

File: docs/proxies.md [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+# Proxies
+
+## Implementing Proxy contracts on Oasis Sapphire
+
+As a confidential Ethereum Virtual Machine (EVM), Oasis prevents external
+access to contract storage or runtime states in order to keep your secrets
+private. This unique feature affects how developers interact with and manage
+smart contracts, particularly when using common Ethereum development tools.
+
+**What are Upgradable Contracts?**
```
Comment (matevz) [inline docs/proxies.md:14]: I would use the sub-sub heading for this.

```suggestion
### What are Upgradable Contracts?
```

File: docs/proxies.md [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+
+Upgradable contracts are smart contracts designed to allow developers to update 
+functionality even after being deployed to a blockchain. This is particularly
+useful for fixing bugs or adding new features without losing the existing state
+or having to deploy a new contract. Upgradability is achieved through proxy
+patterns, where a proxy contract directs calls to an underlying logic contract 
+which developers can swap out without affecting the state stored in the proxy.
+
+**[EIP-1822]: Universal Upgradeable Proxy Standard (UUPS)**
```
Comment (matevz) [inline docs/proxies.md:23]:
```suggestion
#### [EIP-1822]: Universal Upgradeable Proxy Standard (UUPS)
```

File: docs/proxies.md [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+
+EIP-1822 introduces a method for creating upgradable contracts using a proxy
+pattern and specifies a mechanism where the proxy contract itself contains the
+upgrade logic. This design reduces the complexity and potential for errors
+compared to other proxy patterns because it consolidates upgrade functionality
+within the proxy and eliminates the need for additional external management.
+
+**[EIP-1967]: Standard Proxy Storage Slots**
```
Comment (matevz) [inline docs/proxies.md:31]:
```suggestion
#### [EIP-1967]: Standard Proxy Storage Slots
```

**Review by matevz — CHANGES_REQUESTED** [review]

**Review by matevz — COMMENTED** [review]
Do you think we should rather rename the chapter to `upgradable.md`? I can't get rid of a feeling that the chapter is about setting up oracles or paymasters or proxy nodes at first glance when I see the "Proxies" title. The "Upgradable Contracts" is the title I think it rings the right bell for firstcomers.

**Comment by aefhm** [comment]
> Do you think we should rather rename the chapter to `upgradable.md`? I can't get rid of a feeling that the chapter is about setting up oracles or paymasters or proxy nodes at first glance when I see the "Proxies" title. The "Upgradable Contracts" is the title I think it rings the right bell for firstcomers.

I kind of want to 🤔. I hear you on the notion of proxy nodes. My concern with "Upgradeable Contracts" would be that the deterministic deploys are "Clones" so to speak with regards to OpenZeppelin [terminology](https://docs.openzeppelin.com/contracts/4.x/api/proxy). Perhaps "Deployment Customization" or "Upgradable and Deterministic Contracts?"

**Review by matevz — APPROVED** [review]
Excellent!

**Pull request merged**